### ISSUE TITLE
[enhancement] Per-Game ReShade Versions

### ISSUE THREAD

**Issue by CartoonFan** [sonic2kk/steamtinkerlaunch#643]
## System Information
- SteamTinkerLaunch version: v11.12.20221028-1
- Distribution: Arch Linux

## Feature Description

Hey there! I just wanted to suggest two enhancements:

- Per-game Reshade versions
- Shader support for native Linux games

 Sometimes, a different Reshade version is needed as a temporary workaround, but the Reshade version can only be set globally, which makes switching back and forth between versions more tedious than it needs to be, IMO.
 
Also, it might not be something available at present, but it'd be great to have shader support for native Linux games! Currently, it doesn't seem like enabling either vkBasalt (which isn't being developed right now, as I understand) or Reshade actually does anything for native games.
 
 Thanks!

**Comment by sonic2kk** [comment]
Hi there, thanks for the suggestions!

For per-game ReShade versions, we could just add an override drop-down per-game in the Game Menu. The default drop-down value would be something like "global" and then you can change it to use a custom version.

I'll need to look into what STL currently does to handle downloading ReShade, but if we keep multiple versions installed and just switch based on which version is selected, this might be fairly straightforward. I'm not against adding this, but no ETA on when I'll add it. It'll need a little investigation beforehand :-)

As for shader support in native Linux games, ReShade does not support Linux games and only supports Windows games, so it is not possible to add ReShade support for native Linux games. Sorry about that :(

VkBasalt recently did stop development, though I'm curious as to why it's not working for you. Are you sure the games you are testing are using Vulkan? You can check the rendering backend by enabling MangoHud, it'll tell you if it's using DXVK, vkd3d-proton, native Vulkan, etc etc. If it still doesn't work I'll have to investigate this as a separate issue.

If you know of any other shader utilities for Linux please let me know and I'll see about adding support for them! :-)

**Comment by CartoonFan** [comment]
> Hi there, thanks for the suggestions!

Thanks for replying back, too :grin: 

> For per-game ReShade versions, we could just add an override drop-down per-game in the Game Menu. The default drop-down value would be something like "global" and then you can change it to use a custom version.
> 
> I'll need to look into what STL currently does to handle downloading ReShade, but if we keep multiple versions installed and just switch based on which version is selected, this might be fairly straightforward. I'm not against adding this, but no ETA on when I'll add it. It'll need a little investigation beforehand :-)

Awesome :+1: 

> As for shader support in native Linux games, ReShade does not support Linux games and only supports Windows games, so it is not possible to add ReShade support for native Linux games. Sorry about that :(
> 
> VkBasalt recently did stop development, though I'm curious as to why it's not working for you. Are you sure the games you are testing are using Vulkan? You can check the rendering backend by enabling MangoHud, it'll tell you if it's using DXVK, vkd3d-proton, native Vulkan, etc etc. If it still doesn't work I'll have to investigate this as a separate issue.

So I did a bit more digging, and it seems that the vkBasalt config needed to be manually created beforehand. Copied one from the vkBasalt repo, changed some settings, and gave it another try. Things seem to be working mostly OK for now, although I couldn't get LUTs to work properly :thinking: . I haven't tried using the Reshade shaders with it yet, either. I'm testing with Axiom Verge, which seems to have native Vulkan support, so...that part shouldn't be a problem :smile: I'd suggest maybe making the config installation process a little more user-friendly (or at least more transparent), if you have room on your plate for that.

> If you know of any other shader utilities for Linux please let me know and I'll see about adding support for them! :-)

Will do! I don't know of anything else right now, either, but...who knows? Maybe someone will come up with an **awesome** shader program for Linux/Dosbox/ScummVM/etc.

Thanks again!

**Comment by sonic2kk** [comment]
> So I did a bit more digging, and it seems that the vkBasalt config needed to be manually created beforehand.

Yup, this is correct. Glad you got it working!

>  Things seem to be working mostly OK for now, although I couldn't get LUTs to work properly

This is not a SteamTinkerLaunch problem, but a VkBasalt problem. SteamTinkerLaunch doesn't develop, install or configure VkBasalt, it just lets you enable it from a UI so you don't have to remember which commands to set :slightly_smiling_face: 

> I'd suggest maybe making the config installation process a little more user-friendly 

We don't develop, install or configure VkBasalt, we just let users who have it installed enable it more easily, so this isn't desirable or planned I'm afraid. The tools you enable and use with SteamTinkerLaunch are intended to be tools you understand, we only install the Windows tools to make it easier to manage with Wine. You're still expected to know how to use those tools though, and the same goes for native tools.

> Maybe someone will come up with an awesome shader program for Linux/Dosbox/ScummVM/etc.

That would be pretty neat. I mean, this is how even recent tools like DXVK came to be so we can hope :smile: 

<hr>

Going to leave this issue open since the request to add per-game ReShade versions is valid and something I would like to see implemented (either by me or by another contributor).

**Comment by sonic2kk** [comment]
From having a quick skim it seems like STL downloads (and maybe extracts?) the specified ReShade version to `$HOME/.config/steamtinkerlaunch/reshade`, and doesn't download/store multiple ReShade versions. There might be some tweaking needed here. Not sure how much effort will be required for it just yet though.

**Comment by CartoonFan** [comment]
Nice. I just wanted to comment a bit more on your previous statements:

> > Things seem to be working mostly OK for now, although I couldn't get LUTs to work properly
> 
> This is not a SteamTinkerLaunch problem, but a VkBasalt problem. SteamTinkerLaunch doesn't develop, install or configure VkBasalt, it just lets you enable it from a UI so you don't have to remember which commands to set slightly_smiling_face

That's fair. I was just wondering if there was a simple fix you knew of :smile: 

> > So I did a bit more digging, and it seems that the vkBasalt config needed to be manually created beforehand.
> 
> Yup, this is correct. Glad you got it working!
> 
> > I'd suggest maybe making the config installation process a little more user-friendly
> 
> We don't develop, install or configure VkBasalt, we just let users who have it installed enable it more easily, so this isn't desirable or planned I'm afraid. The tools you enable and use with SteamTinkerLaunch are intended to be tools you understand, we only install the Windows tools to make it easier to manage with Wine. You're still expected to know how to use those tools though, and the same goes for native tools.

Again, this is understandable. Would it be at least possible to drop a default config in the game directory with the correct shader/texture paths? Maybe like a "vkBasalt install" option? I'm no shell expert, but I feel that sed and cp could pull that off.

I'm not trying to impose, and I just want to emphasize that I'm not trying to have you make a GUI or anything. I'm just trying to spare someone else from the same confusion I experienced :pray:

Thanks!

**Comment by sonic2kk** [comment]
> I'm no shell expert, but I feel that sed and cp could pull that off.

They could, but this isn't up to SteamTinkerLaunch imo. It doesn't configure your tools for you.

> I'm just trying to spare someone else from the same confusion I experienced 

I get that, and I'm not trying to be rude here but that's user error. If you had used vkBasalt as a launch option you would've experienced the same problem. Reading up on how to use VkBasalt (like you did) is how a user can save themselves from that confusion.

Hope that justifies the position a little :-)

**Comment by CartoonFan** [comment]
> They could, but this isn't up to SteamTinkerLaunch imo. It doesn't configure your tools for you.

Sorry, but...I don't really get this one? There are a lot of config options for several tools featured in SteamTinkerLaunch :confused:

> I get that, and I'm not trying to be rude here but that's user error. If you had used vkBasalt as a launch option you would've experienced the same problem. Reading up on how to use VkBasalt (like you did) is how a user can save themselves from that confusion.
>
 > Hope that justifies the position a little :-)
 
 Um...OK. Can I at least ask for the wiki to be updated? :sweat_smile:
 
 The only section there says
 
 > ENABLE_VKBASALT: set ENABLE_VKBASALT to 1 to start the game with vkbasalt (does nothing in steamtinkerlaunch itself, but just exports the upstream variable)

which makes it seem like all one has to do is set ENABLE_VKBASALT to 1 in order for vkBasalt to work.

I'm not trying to be rude or difficult either :pray:

**Comment by sonic2kk** [comment]
The wiki could be updated but again, a user is expected to know how to use the tools they enable with SteamTinkerLaunch. You're expected to know how to use your tools before you configure them with SteamTinkerLaunch.

It is assumed that a user will already be using VkBasalt and other similar tools before enabling them in SteamTinkerLaunch.

The steps listed on the wiki show you how to use VkBasalt with SteamTinkerLaunch. The steps with the configuration file are not specific to using VkBasalt with SteamTinkerLaunch, it's something you would have to do regardless. Reading the information on the tool's repo beforehand, which the wiki page should link to,  should be enough to a user to get setup.

<hr>

> There are a lot of config options for several tools featured in SteamTinkerLaunch

Sorry I didn't see this edit. Could you explain a little more? Maybe on my end I didn't explain it very well.

SteamTinkerLaunch does not, for example, create a GameMode configuration file. It also does not install tools like GameMode or MangoHud. It  doesn't set those tools up for you, it's expected that you've already correctly installed and set up your tools before you use them with SteamTinkerLaunch. It will allow you to do some configuration like enabling and disabling, a GUI to pass arguments for the likes of MangoHud, but this isn't applicable to VkBasalt. If you weren't using it before you started using SteamTinkerLaunch, you should refer to a tool's documentation primarily for how to get through the initial setup.

<hr>

Note this section of the wiki you quoted:

> does nothing in steamtinkerlaunch itself, but just exports the upstream variable

This is primarily what SteamTinkerLaunch does. It gives you a GUI to enable tools. If you had exported the VkBasalt variable from the Steam launch options it wouldn't have worked initially.

**Comment by CartoonFan** [comment]
> > There are a lot of config options for several tools featured in SteamTinkerLaunch
> 
> Sorry I didn't see this edit. Could you explain a little more? Maybe on my end I didn't explain it very well.
> 
> SteamTinkerLaunch does not, for example, create a GameMode configuration file. It also does not install tools like GameMode or MangoHud. It doesn't set those tools up for you, it's expected that you've already correctly installed and set up your tools before you use them with SteamTinkerLaunch. It will allow you to do some configuration like enabling and disabling, a GUI to pass arguments for the likes of MangoHud, but this isn't applicable to VkBasalt. If you weren't using it before you started using SteamTinkerLaunch, you should refer to a tool's documentation primarily for how to get through the initial setup.

After really thinking about what I wrote, I figured that maybe SteamTinkerLaunch doesn't actually touch the config files of the tools themselves; it really only adds flags and variables. So in that sense, things fit together quite well. Unless I'm wrong, in which case feel free to correct me :smile: 

> The wiki could be updated but again, a user is expected to know how to use the tools they enable with SteamTinkerLaunch. You're expected to know how to use your tools before you configure them with SteamTinkerLaunch.
> 
> It is assumed that a user will already be using VkBasalt and other similar tools before enabling them in SteamTinkerLaunch.
> 
> The steps listed on the wiki show you how to use VkBasalt with SteamTinkerLaunch. The steps with the configuration file are not specific to using VkBasalt with SteamTinkerLaunch, it's something you would have to do regardless. Reading the information on the tool's repo beforehand, which the wiki page should link to, should be enough to a user to get setup.
>

> Note this section of the wiki you quoted:
> 
> > does nothing in steamtinkerlaunch itself, but just exports the upstream variable
> 
> This is primarily what SteamTinkerLaunch does. It gives you a GUI to enable tools. If you had exported the VkBasalt variable from the Steam launch options it wouldn't have worked initially.

So...I think I'm getting what you're trying to say now: It's up to me or @DadSchoorse to have the config file where it needs to be :sweat_smile:

It's just a bit frustrating that vkBasalt won't work **at all** without a config file already present. It could at least generate a config file or something on the first run if there are no real "defaults" that would make sense.

Anyway, that's on us, not you. Thanks for working with me :+1:

In my defense, I've only been using SteamTinkerLaunch for a few days, and I only really put the effort to work with vkBasalt today :laughing:

**Comment by sonic2kk** [comment]
> So...I think I'm getting what you're trying to say now: It's up to me or [maintainer] to have the config file where it needs to be

Yup, that's basically the idea. For initial setup ideally the program itself, or the packager, would do this. But I'm not familiar with the project and I appreciate that development upstream (or lack thereof) may play a role here.

This is configuration error on the part of VkBasalt though, not of SteamTinkerLaunch, which is why I don't think it should be something that we handle. This configuration problem happens outside of SteamTinkerLaunch, I don't think it's something we should "work around" so to speak.

For tools like VkBasalt but also other native tools like MangoHud or GameMode, which are applied in a similar way to VkBasalt: If a user wasn't using them before SteamTinkerLaunch, they should learn more about it first. Hope that makes sense.

<hr>

No problem on discussing it. I like getting discussion on issues. I'm sorry it didn't get the resolution you were hoping for, but I think that if anything this is better handled upstream or by the packager.

I was thinking now of adding an option to point to a custom VkBasalt config file, but I'm slightly conscious of spending too much time improving compatibility with a program where upstream development may not be so active. However having said that, PRs relating to improving support for VkBasalt support are welcome.

**Comment by CartoonFan** [comment]
> > So...I think I'm getting what you're trying to say now: It's up to me or [maintainer] to have the config file where it needs to be
> 
> Yup, that's basically the idea. For initial setup ideally the program itself, or the packager, would do this. But I'm not familiar with the project and I appreciate that development upstream (or lack thereof) may play a role here.
> 
> This is configuration error on the part of VkBasalt though, not of SteamTinkerLaunch, which is why I don't think it should be something that we handle. This configuration problem happens outside of SteamTinkerLaunch, I don't think it's something we should "work around" so to speak.
> 
> For tools like VkBasalt but also other native tools like MangoHud or GameMode, which are applied in a similar way to VkBasalt: If a user wasn't using them before SteamTinkerLaunch, they should learn more about it first. Hope that makes sense.
> 
> No problem on discussing it. I like getting discussion on issues. I'm sorry it didn't get the resolution you were hoping for, but I think that if anything this is better handled upstream or by the packager.
> 
> I was thinking now of adding an option to point to a custom VkBasalt config file, but I'm slightly conscious of spending too much time improving compatibility with a program where upstream development may not be so active. However having said that, PRs relating to improving support for VkBasalt support are welcome.

Totally understand. Thank you :+1:

**Comment by zany130** [comment]
this was actually on my todo list of enhancements to work on for a while  (never got around to reporting it) but the main thing I wanted to do is add support for the addon version of reshade which lets you inject varaouis addons into your games ( for example there is one to toggle in -game shaders on and off so you can disable things like HUDS or even different textures of game objects if you wanted to)

the only problem is that this version of reshade is not anti-cheat compatible and WILL get you flagged/banned

and with gameconqurer/CE being removed recently for similar concerns https://github.com/sonic2kk/steamtinkerlaunch/issues/618
IDK if it would be a good idea to have an option to download the addon enabled version.

maybe we could do a game specific override for the reshade version ( I think the variable was `RSVER` or something like that so we could have `OVRSVER` or something like that)

and then the user will just type in the version they want into the text box in game menu, so for example `5.4.2_Addon` for the addon version

this would work because iirc reshade just looks at what `RSVER` is and adds that to the end of the reshade url `https://reshade.me/downloads/$RSVER` and extracts/installs that version


**TLDR**
________

all we would have to do to implement this is 


1. add a override variable that takes in the reshade version (and `_Addon` if they want that or we can add a check box which does that for the user, but  that will bring up the same issues as the CE/gameconqurer issue)
2. set `RSVER` to `RSVER_OVER`(i think thats better)
when `RSVER_OVER` has a value.


with those two things **I think** we would be able to complete this request.

I wouldn't mind taking a stab a this if this is correct/desired though don't know when  😅 (though i expect it to be pretty straight forward to do

**Comment by sonic2kk** [comment]
Awesome! You seem to know a lot more about ReShade than me so I would **very** much appreciate if you had time to look into this! :heart: No rush or anything on getting it implemented but thank you for stepping up and offering to add this. 

I agree that having an `RSVER_OVER` var or something along those lines is a good idea (seems to be convention to have all upcase, no punctuation variable names in STL - that's nitpicky though :stuck_out_tongue:). This would store the Per-Game ReShade version override, correct?

I also agree with your concerns with the ReShade AddOn extension(?). I appreciate the thought but I think it's best if we don't include it for the reasons you outlined. I am not really a fan of bending over for invasive anti-cheat, but I don't want users getting banned for using SteamTinkerLaunch.

My question here is, when STL downloads ReShade, does it download it into its own versioned folder? We wouldn't want to have to fetch and re-download various ReShade versions each time, it would be nice if these could be stored and if STL would just load/fetch the version chosen by the user. If it's not currently the case and if it's not too much trouble, could you take a look at that also?

I think on the existing option Global Menu and the new option Game Menu should be a textbox with a dropdown, where a user can enter their own ReShade version or choose from a list of dropdowns. On the Game Menu, the default should stay as I think `5.0.0` currently. But if there are "common" versions of ReShade that someone might want to use instead, we could include that in the dropdown. On the Game Menu, the default ReShade version (which would correspond to `RSVER_OVER`) should be `global` and the dropdown then would list other common ReShade versions that a user might want to roll back to, just like on the Global Menu but this makes it more accessible for per-game customisation.

That means if a user knows that, for example, `4.2.0` is more compatible with their game than the global version, they can just select it from the dropdown. Then if the ReShade version is just set as `global` we'll check the `RSVER` variable and use that as our ReShade version.

I hope that makes sense, given your input here this is the solution I'm envisioning to this problem. But I could be talking nonsense here or there could be pitfalls with this approach, so **please** give your thoughts here. @CartoonFan you also are welcome to chip into this discussion of course if anything comes to mind. I don't use ReShade so the implementation outline I've given might not make any sense.

**Comment by zany130** [comment]
> I agree that having an RSVER_OVER var or something along those lines is a good idea (seems to be convention to have all upcase, no punctuation variable names in STL - that's nitpicky though stuck_out_tongue). This would store the Per-Game ReShade version override, correct?

Maybe `RSVERTEM` or `TEMRSVER` then might fit better?  and  yes it would store the per-game ReShade version

>I also agree with your concerns with the ReShade AddOn extension(?). I appreciate the thought but I think it's best if we don't include it for the reasons you outlined. I am not really a fan of bending over for invasive anti-cheat, but I don't want users getting banned for using SteamTinkerLaunch.

no worries I think it would also keep the code simpler if we don't have a check box that will just essentially append `_Addon` the current `RSVER` (btw if it wasn't clear, both are hosted on the same site, and to get the Addon version you would just `_Addon` to the version number in the URL for example `5.4.2_Addon` so https://reshade.me/downloads/ReShade_Setup_5.4.2.exe for normal and https://reshade.me/downloads/ReShade_Setup_5.4.2_Addon.exe for addon ) Technically by having this as a text box in the interface the user can add `_Addon` to the end of the version if they so wish. I don't know if you would like this to be documented in the wiki or not. Personally, I don't see the issue with it being in the wiki with a warning, but I agree having it as a check box or even a dropdown option is making it too easy for uniformed users to enable it.

>My question here is, when STL downloads ReShade, does it download it into its own versioned folder? We wouldn't want to have to fetch and re-download various ReShade versions each time, it would be nice if these could be stored and if STL would just load/fetch the version chosen by the user. If it's not currently the case and if it's not too much trouble, could you take a look at that also?

I think this will just work as it take a look at my reshade folder 
![image](https://user-images.githubusercontent.com/25832539/198851910-11c78ba2-fd8e-4fd8-b712-3edda079a6b4.png)

I will need to confirm this but the way STL handles reshade right now is
1.it will download the exe from `https://reshade.me/downloads/ReShade_Setup_$RSVER.exe` into  the stl downloads/reshade folder
2.it will extract Reshade32.and 64.dll (And Reshade32 and 64.json for Vulkan games)
3. It then will copy the correct files into the games folder (depending on if its directx or vulkan and 64 or 32bit(


so as it is now, it will keep the different exe versions (thus shouldn't need to download it if it already exists but I would need to check that), but the .dlls and .json files will always be overwritten (maybe we should tell stl to extract it into a folder named after `RSVER`?)

**Comment by sonic2kk** [comment]
>  Personally, I don't see the issue with it being in the wiki with a warning, but I agree having it as a check box or even a dropdown option is making it too easy for uniformed users to enable it.

Ah okay I understand you now. Sorry, you can probably tell I really don't know much about ReShade :sweat_smile: So if I understand correctly, it would really just be the ability to parse and install the `_Addon` ReShade executable, right? If so, we could add a warning about this on the wiki noting that it can work but will get you banned from games. If I'm understanding correctly here, I think that's the best approach.

> I think this will just work as it take a look at my reshade folder

Ahhhhh okay, so each executable is downloaded and we extract the DLLs for each version to the folder (and then use those to setup ReShade for the game). As long as the names of the files don't change between ReShade versions (which I think they do stay the same) then I think this will work just fine. If there are name changes we could manually check based on the ReShade version but that's probably not going to be a problem :-)

> so as it is now, it will keep the different exe versions (thus shouldn't need to download it if it already exists but I would need to check that), but the .dlls and .json files will always be overwritten (maybe we should tell stl to extract it into a folder named after RSVER?)

Extracting into a folder might be nice to keep it separated but If getting those DLLs is quick, it's probably okay to just keep it all in the top level ReShade folder and have those DLLs and JSON files get overwritten. "Quick" is subjective, but the Steam Deck could be a good benchmark here. If it's significantly faster to keep the files instead of extracting them each time it might be best to have separate folders for each ReShade version with the files extracted and ready. I guess it gets those files from running the executable, right? Running that with Proton/Wine might add a bit of overhead on each run so I could see it being noticeably faster to have the files extracted in a folder for each ReShade version. But if that's too much work, or if it doesn't take too long to extract the ReShade files each time, we could just leave it as-is :smiley: 

Thanks for digging into this, it gave some really good insight :+1:

**Comment by zany130** [comment]
> Extracting into a folder might be nice to keep it separated but If getting those DLLs is quick, it's probably okay to just keep it all in the top level ReShade folder and have those DLLs and JSON files get overwritten. "Quick" is subjective, but the Steam Deck could be a good benchmark here. If it's significantly faster to keep the files instead of extracting them each time it might be best to have separate folders for each ReShade version with the files extracted and ready. I guess it gets those files from running the executable, right? Running that with Proton/Wine might add a bit of overhead on each run so I could see it being noticeably faster to have the files extracted in a folder for each ReShade version. But if that's too much work, or if it doesn't take too long to extract the ReShade files each time, we could just leave it as-is smiley

it just runs `unzip` on the exe see https://github.com/zany130/steamtinkerlaunch/blob/b361e6181628ab56296f7cc33108abcf473ef262/steamtinkerlaunch#L7425

also opened a PR with preliminary work on this https://github.com/sonic2kk/steamtinkerlaunch/pull/645

**Comment by sonic2kk** [comment]
Hey @CartoonFan, this feature was implemented by @zany130 with 46620cf3aa4675ecbbe8cff4ac9028bb786ab078 - This issue has now been closed as the feature is implemented :smile:

**Comment by CartoonFan** [comment]
Thanks! I'll gladly try it out!

**Comment by zany130** [comment]
finally got around to updating the reshade wiki page @CartoonFan let me know if you think anything can be made clearer. And of course, feel free to let me know of any bugs or anything that can be improved

**Comment by CartoonFan** [comment]
> finally got around to updating the reshade wiki page @CartoonFan let me know if you think anything can be made clearer. And of course, feel free to let me know of any bugs or anything that can be improved

Thanks! I'll take a look :grin:
